### PR TITLE
Experiment with attribute sourcing

### PR DESCRIPTION
**Pull request by ockham**
Here's a very bare-bones, WIP, implementation of attribute sourcing.

For more background, refer to @luisherranz' great video in #27.

Feel free to pick it up and play with it; otherwise, I'll continue working on it on Friday 🙂 

image

### CODE REVIEW

File: src/gutenberg-packages/frontend.js [hunk 1]
```diff
@@ -59,6 +59,14 @@ class GutenbergBlock extends HTMLElement {
 			const attributes = JSON.parse(
 				this.getAttribute( 'data-gutenberg-attributes' ),
 			);
+			const sourcedAttributes = JSON.parse(
+				this.getAttribute( 'data-gutenberg-sourced-attributes' ),
+			)
+
+			for ( const attr in sourcedAttributes ) {
+				attributes[ attr ] = matcherFromSource( sourcedAttributes[ attr ] )( this );
```
Comment (ockham) [inline src/gutenberg-packages/frontend.js:67]: `matcherFromSource` is a wrapper around `hpq`, which expects a DOM element tree. Good thing that `this` is one! 😄

File: src/gutenberg-packages/utils.js [hunk 2]
```diff
@@ -45,3 +70,12 @@ export const pickKeys = ( obj, arr ) => {
 	}
 	return result;
 };
+
+// See https://github.com/WordPress/gutenberg/blob/trunk/packages/blocks/src/api/parser/get-block-attributes.js#L185
+export const matcherFromSource = memoize( ( sourceConfig ) => {
```
[the remark this answers is not in the document]
Comment (ockham) [reply src/gutenberg-packages/utils.js]: b80e256

**Comment by ockham** [comment]
I was originally planning to add support for more attribute sources [here](https://github.com/WordPress/block-hydration-experiments/pull/30/files#diff-1044b2975045e560c8a2975481d87d723ce3d8f069d6e7e5c0e465f5dfff9c7fR76) (`html`, `query`, `meta`, etc -- [see](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-attributes/#value-source)), but now I'm thinking it might not add much value to the proof-of-concept (and just amount to copying [code over from GB](https://github.com/WordPress/gutenberg/blob/c0e618a5947c7ff1a1e88d7ab6d9e0749ce31771/packages/blocks/src/api/parser/get-block-attributes.js#L185-L225)).

So maybe this is ready for review/merging?

**Review by luisherranz — APPROVED** [review]
Sure, approved so you can merge as you please 🙂

However, would you mind adding a non-sourced attribute before you do so? Just a simple one can do. _If not, the demo won't ship any non-sourced attribute anymore._

**Comment by ockham** [comment]
> Sure, approved so you can merge as you please 🙂
> 
> However, would you mind adding a non-sourced attribute before you do so? Just a simple one can do. _If not, the demo won't ship any non-sourced attribute anymore._

Thanks, good point! I made the counter value an attribute (i.e. it's not possible to set the initial counter value from the editor). (Changing it on the frontend won't persist it, though.)

**Comment by luisherranz** [comment]
Awesome. Thank you, Bernie 🙂

**Pull request merged**